pkg/koordlet/metricsadvisor/collectors/prom/prom_collector.go

```diff
@@ -0,0 +1,146 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prom
+
+import (
+	"fmt"
+	"time"
+
+	"go.uber.org/atomic"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/metriccache"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/metricsadvisor/framework"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/statesinformer"
+)
+
+const CollectorName = "PromCollector"
+
+// The collector is for collecting the metrics (prometheus format encoded) from external prometheus exporters.
+// To simplify the implementation of the prometheus metrics collection, the collector allows multiple scrapers to
+// register, and a scraper is responsible for scrape and decode the metrics from a prometheus exporter.
+// we expect a collector collects from only one exporter
+// and decode the raw metrics into the metrics classified by the metric level (node, pod, container), and we will
+// enhance the prometheus metrics collecting framework based on the metric advisor framework.
+type collector struct {
+	appendableDB    metriccache.Appendable
+	statesInformer  statesinformer.StatesInformer
+	scrapers        []Scraper
+	collectInterval time.Duration
+	rulePath        string
+	started         *atomic.Bool
+}
+
+func New(opt *framework.Options) framework.Collector {
+	return &collector{
+		appendableDB:    opt.MetricCache,
+		statesInformer:  opt.StatesInformer,
+		collectInterval: opt.Config.CollectPromMetricInterval,
+		rulePath:        opt.Config.CollectPromMetricRulePath,
+		started:         atomic.NewBool(false),
+	}
+}
+
+func (c *collector) Enabled() bool {
+	return c.collectInterval > 0
+}
+
+func (c *collector) Setup(s *framework.Context) {}
+
+func (c *collector) Run(stopCh <-chan struct{}) {
+	if !cache.WaitForCacheSync(stopCh, c.statesInformer.HasSynced) {
+		// exit when statesInformer sync failed
+		klog.Fatalf("timed out waiting for states informer caches to sync")
+	}
+
+	if err := c.initScrapers(); err != nil {
+		klog.Warningf("failed to init scrapers, err: %s", err)
+		return
+	}
+
+	go wait.Until(c.runOnce, c.collectInterval, stopCh)
+}
+
+func (c *collector) Started() bool {
+	return c.started.Load()
+}
+
+func (c *collector) initScrapers() error {
+	// load rules from the rule
+	rule, err := LoadCollectRuleFromFile(c.rulePath)
+	if err != nil {
+		return fmt.Errorf("load collect rule failed, file %s, err: %w", c.rulePath, err)
+	}
+	// init scrapers
+	var scrapers []Scraper
+	configs, err := rule.ToConfigs()
+	if err != nil {
+		return fmt.Errorf("convert collect rule to config failed, rule %+v, err: %w", rule, err)
```

Comment:
these errors should be exposed as event or metrics

Reply:
added